kernel/src/process/posix_thread/exit.rs

```diff
@@ -1,66 +1,100 @@
+};
 use crate::{
     current_userspace,
     prelude::*,
-    process::{do_exit_group, TermStatus},
-    thread::{Thread, Tid},
+    process::{exit::exit_process, TermStatus},
+    thread::AsThread,
 };
 
-/// Exits the thread if the thread is a POSIX thread.
+/// Exits the current POSIX thread.
 ///
 /// # Panics
 ///
-/// If the thread is not a POSIX thread, this method will panic.
-pub fn do_exit(thread: &Thread, posix_thread: &PosixThread, term_status: TermStatus) -> Result<()> {
-    if thread.is_exited() {
-        return Ok(());
+/// If the current thread is not a POSIX thread, this method will panic.
+pub fn do_exit(term_status: Option<TermStatus>) {
+    let current_task = Task::current().unwrap();
+    let current_thread = current_task.as_thread().unwrap();
+
+    // We should only change the thread status when running as the thread, so no race conditions
+    // can occur in between.
+    if current_thread.is_exited() {
+        return;
     }
-    thread.exit();
-
-    let tid = posix_thread.tid;
-
-    let mut clear_ctid = posix_thread.clear_child_tid().lock();
-    // If clear_ctid !=0 ,do a futex wake and write zero to the clear_ctid addr.
-    if *clear_ctid != 0 {
-        // FIXME: the correct write length?
-        if let Err(e) = current_userspace!().write_val(*clear_ctid, &0u32) {
-            debug!("Ignore error during exit process: {:?}", e);
-        }
-        futex_wake(*clear_ctid, 1, None)?;
-        *clear_ctid = 0;
+    current_thread.exit();
+
+    let current_thread = current_thread.as_posix_thread().unwrap();
+    let current_process = current_thread.process();
+
+    // According to Linux's behavior, the last thread's exit code will become the process's exit
+    // code, so here we should just overwrite the old value (if any).
+    if let Some(term_status) = term_status {
+        current_process.status().set_exit_code(term_status.as_u32());
     }
-    drop(clear_ctid);
-    // exit the robust list: walk the robust list; mark futex words as dead and do futex wake
-    wake_robust_list(posix_thread, tid);
-
-    if tid != posix_thread.process().pid() {
-        // We don't remove main thread.
-        // The main thread is removed when the process is reaped.
-        thread_table::remove_thread(tid);
+
+    let is_last_thread = {
+        let mut tasks = current_process.tasks().lock();
+        let current_index = tasks
+            .iter()
+            .position(|task| core::ptr::eq(task.as_ref(), current_task.as_ref()))
+            .unwrap();
+        tasks.swap_remove(current_index);
+
+        tasks.is_empty()
+    };
+
+    wake_clear_ctid(current_thread);
+
+    wake_robust_list(current_thread);
+
+    // FIXME: According to Linux behavior, it seems that we shouldn't remove the main thread unless
+    // the process is reaped by its parent. However, this makes it harder to track whether the
+    // exiting thread is the last thread in the process.
+    thread_table::remove_thread(current_thread.tid());
```

Comment:
> But I can't understand why the thread needs to be removed from the thread table. This is a global table, and its contents shouldn't be affected by whether it's the last thread.

My understanding is that for all threads _other than the main thread_, once the thread exits, it can be automatically reaped/recycled and the associated resources can then be freed.
 - So I assume that _all_ references to the thread need to be removed, including those in the global thread table.

As an example, if you exit a non-main thread, you cannot find the thread anymore in `/proc/[self]/task/`:
<details>
<summary> code </summary>

```c
#include <unistd.h>
#include <pthread.h>
#include <sys/syscall.h>
#include <stdlib.h>

static void *thread_main(void *arg)
{
	sleep(1);

	syscall(SYS_exit, 0);
	__builtin_unreachable();
}

int main(void)
{
	pthread_t tid;
	pthread_create(&tid, NULL, &thread_main, NULL);

	system("ls /proc/$PPID/task");
	sleep(3);
	system("ls /proc/$PPID/task");

	syscall(SYS_exit, 0);
	__builtin_unreachable();
}
```
```
> ./a.out
11545  11546
11545
```
</details>

However, if you exit the main thread, the thread can still be found in `/proc/[self]/task` (but `stat` will indicate it as a zombie thread):

<details>
<summary> code </summary>

```c
#include <unistd.h>
#include <pthread.h>
#include <sys/syscall.h>
#include <stdlib.h>

static void *thread_main(void *arg)
{
	system("ls /proc/$PPID/task");
	sleep(3);
	system("ls /proc/$PPID/task");

	syscall(SYS_exit, 0);
	__builtin_unreachable();
}

int main(void)
{
	pthread_t tid;
	pthread_create(&tid, NULL, &thread_main, NULL);

	sleep(1);

	syscall(SYS_exit, 0);
	__builtin_unreachable();
}
```
```
> ./a.out
11730  11731
11730  11731
```
</details>

Reply:
Sorry for the confusion caused by my original comments. What I meant to convey is that the _main thread_ should not be removed from the thread table when doing exit. The main thread should only be removed when the parent process reaps the process. However, the current code appears to remove the main thread from the thread table during the exit process as well.

Reply:
> The main thread should only be removed when the parent process reaps the process. However, the current code appears to remove the main thread from the thread table during the exit process as well.

Well, as I said in the comments, this makes it harder to keep track of whether the exiting thread is the last thread in the process.

I've managed to do this by adding a new type `TaskSet`, in a new commit for easier review. Could you please take a look at this and see if there's a better way to achieve this?

Reply:
> Well, as I said in the comments, this makes it harder to keep track of whether the exiting thread is the last thread in the process.

Why doesn't the original code (referring to the code in the first commit) keep track of which thread is the last one?

If a thread calls `do_exit`, it will remove itself from the process's tasks vector after cleaning up. If there are no more tasks left in the vector, that thread should be the last one.

If a thread calls `do_exit_group`, it marks the process as undergoing an exit group. Consequently, all other threads will perform exit operations and clean themselves up (no other thread can call do_exit_group again). Once all other threads have finished, the thread that called `do_exit_group` will finally clear the process. Therefore, the thread that invokes `do_exit_group` should be the last thread.

It seems that whether a thread is removed from the thread_table is unrelated to determining the last thread. Am I misunderstanding something?

Reply:
We don't want to remove the main thread from the global thread table. I think this implies that we don't want to remove it from the process thread table either (i.e. the `tasks` field in `Process`).
- But if the main thread cannot be removed from the `tasks` field, it will be harder to keep track of whether the exiting thread is the last thread in the process (since the main thread _may_ exit before, but it can still be in the `tasks` field).

There are several reasons for the aforementioned "implication":
 - It's Linux behavior. The main thread becomes a zombie thread that can still be found in `/proc/[pid]/tasks`. If we remove it from the process thread table, how can we find it when looking at `/proc/[pid]/tasks`?
   - Theoretically, we can work around this by manually adding the zombie main thread back when generating `procfs` content, but this is quite error-prone.
      - Remember that the main thread may or may not have been a zombie.
      - Remember that we need to provide not only the `/proc/[pid]/tasks` directory, but also some metadata (such as the number of threads in `/proc/[pid]/status`).
 - It's more consistent behavior. The main thread becomes a zombie thread, but it's still there, so it should not be removed from the process thread table.
 - More importantly, if we remove it from the process thread table, but don't remove it from the global thread table, how can the resources of this thread be released when the parent process kills the zombie process? After all, it cannot find the thread in the process thread table, so there may be resource leakage.
   - See the code below. Again, we can work around this by manually freeing the resources of the zombie main thread, but it's tedious and error-prone.

https://github.com/asterinas/asterinas/blob/5e35704e3896897550ac4ee383923d1e054f3dd1/kernel/src/process/wait.rs#L94-L96

So I think it's better to improve the maintenance of `tasks` in `process` to achieve the following at the same time:
 - It does not allow removing the zombie main thread.
 - But it can keep track of whether the exiting thread is the last thread.

This is done by the new type `TaskSet`, which consists not only of `Vec<Arc<Task>>`, but also of a boolean flag to indicate whether the main thread becomes a zombie.
 - In addition, it brings more benefits by putting the `has_exited_group` flag also in the `TaskSet` to avoid manually preventing race conditions when `clone` and `exit_group` are called at the same time (in case that a new task is created but does not receive the SIGKILL signal, messing up the process state).

Reply:
Make sense to me!

I previously thought that the main thread could be removed from a process's tasks and kept in the global table. However, as you mentioned, this approach would complicate maintaining the process state. Therefore, I think your current design is better.